presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.DataSize;
+import org.weakref.jmx.Managed;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import static com.facebook.presto.operator.Operator.NOT_BLOCKED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MemoryPool
+{
+    private final MemoryPoolId id;
+    private final long maxBytes;
+    private final boolean enableBlocking;
+
+    @GuardedBy("this")
+    private long freeBytes;
+
+    @Nullable
+    @GuardedBy("this")
+    private SettableFuture<?> future;
+
+    public MemoryPool(MemoryPoolId id, DataSize size, boolean enableBlocking)
+    {
+        this.id = requireNonNull(id, "name is null");
+        maxBytes = size.toBytes();
+        this.enableBlocking = enableBlocking;
+        freeBytes = size.toBytes();
+    }
+
+    public MemoryPoolId getId()
+    {
+        return id;
+    }
+
+    public synchronized MemoryPoolInfo getInfo()
+    {
+        return new MemoryPoolInfo(id, maxBytes, freeBytes);
+    }
+
+    /**
+     * Reserves the given number of bytes. Caller should wait on the returned future, before allocating more memory.
+     */
+    public synchronized ListenableFuture<?> reserve(long bytes)
+    {
+        checkArgument(bytes >= 0, "bytes is negative");
+        freeBytes -= bytes;
+        if (freeBytes <= 0) {
+            if (future == null) {
+                future = SettableFuture.create();
```

Comment:
Discussed offline with @dain, and this is actually fine because we just end up with negative free memory (the pool becomes overcommitted) which is a totally expected situation. In that situation, the pool doesn't unblock anyone until it is no longer over committed (hence the `freeBytes > 0` check in `free()`)